import csv, os

ROOT = "/well/longread/users/akl399/bakeoff"

# Output: ROOT/analysis/de_novo_assembly/data/{acronym}.{build}-{method}/assembly.fa
# Input: (1) mapped reads; (2) region defination

tools = {
	"samtools": "/apps/well/samtools/1.4.1/bin/samtools",
	"flye": "/well/ont/apps/Flye-2.5/bin/flye",
	"racon": "/well/ont/apps/racon/build/bin/racon",
	"minimap2": "/well/ont/apps/minimap2-v2.14/minimap2/minimap2",
	"medaka": "/well/ont/apps/medaka/venv/bin/activate"
}

def load_regions( filename = ROOT + '/resources/regions.tsv' ):
	"""
	Returns a list of regions, each as a OrderedDict.
	Example: 
		OrderedDict([('name', 'T Cell Receptor Alpha'), 
			('acronym', 'TRA'), ('build', 'GRCh38'), 
			('chromosome', 'chr14'), ('start', '21170441'), 
			('end', '22134544'), 
			('start_sequence', 'CCTACCCCACACACTTATCACCCAGCAGGGAACCCTCAGGTTGGGCCCACAGCCCCCATT'), 
			('end_sequence', 'CCCTGGAAACCATTATTTTATTCTCTGCTTCTATGAATTTGACTATTTTAGATACCTCAAATAAGTGTAATCATGCATTATTTG')])
	"""
	regions = []
	with open( filename, newline = '' ) as file:
		regionReader = csv.DictReader( file, delimiter = '\t' )
		for region in regionReader:
			regions.append( dict( region ))
	return regions

regions = load_regions()

def get_region_definition(acronym, build, regions):
	result = [ region for region in regions if region['build'] == "GRCh" + build and region['acronym'] == acronym ]
	if len( result ) != 1:
		raise Exception( "No region with  acronym %s and build %s can be found." % ( acronym, build ) )
	return result[0]

rule Extract_target_region:
	input: 
		bam = "/well/longread/projects/nanopore/data/minimap2_all-build{build}.bam",
		region_definition = ROOT + '/resources/regions.tsv'

	output:
		reads = ROOT + "/analysis/de_novo_assembly/data/Build{build}/{acronym}/Extract_target_region/reads.fastq"
	
	params:
		chromosome = lambda wildcards: get_region_definition(wildcards.acronym, wildcards.build, regions)["chromosome"],
		start = lambda wildcards: get_region_definition(wildcards.acronym, wildcards.build, regions)["start"],
		end = lambda wildcards: get_region_definition(wildcards.acronym, wildcards.build, regions)["end"],
		samtools = tools['samtools'],

	shell:
		"{params.samtools} view -b {input.bam} {params.chromosome}:{params.start}-{params.end} | {params.samtools} fastq - > {output.reads}"

rule Flye:
	input: 
		reads = ROOT + "/analysis/de_novo_assembly/data/Build{build}/{acronym}/Extract_target_region/reads.fastq"

	output: 
		consensus = ROOT + "/analysis/de_novo_assembly/data/Build{build}/{acronym}/{method}/Flye/assembly.fasta",
		symlink = ROOT + "/analysis/de_novo_assembly/data/Build{build}/{acronym}/{method}/Racon_input.symlink.fasta"

	params:
		flye = tools['flye'],
		output_folder = ROOT + "/analysis/de_novo_assembly/data/Build{build}/{acronym}/{method}/Flye/"

	shell: 
		"""
		{params.flye} --nano-raw {input.reads} -g 1m -o {params.output_folder} -t 4 --asm-coverage 40
		ln -s {output.consensus} {output.symlink}
		"""

rule Racon:
	input: 
		scaffold = ROOT + "/analysis/de_novo_assembly/data/Build{build}/{acronym}/{method}/Racon_input.symlink.fasta",
		reads = ROOT + "/analysis/de_novo_assembly/data/Build{build}/{acronym}/Extract_target_region/reads.fastq"

	output:
		scaffold = ROOT + "/analysis/de_novo_assembly/data/Build{build}/{acronym}/{method}/3xRacon/racon-3.ctg.fa.gz",
		symlink = ROOT + "/analysis/de_novo_assembly/data/Build{build}/{acronym}/{method}/Medaka_input.symlink.ctg.fa.gz"

	params:
		racon = tools['racon'],
		minimap2 = tools['minimap2'],
		output_folder = ROOT + "/analysis/de_novo_assembly/data/Build{build}/{acronym}/{method}/3xRacon/"

	shell:
		"""
		SCAFFOLD={input.scaffold}
		READS={input.reads}
		for ROUND in $(seq -w 1 3); do
			echo "Running round ${{ROUND}} consensus..."
    		READS2TIGS={params.output_folder}racon-${{ROUND}}.paf.gz
    		NEWSCAF={params.output_folder}racon-${{ROUND}}.ctg.fa.gz

    		{params.minimap2} -t 4 ${{SCAFFOLD}} ${{READS}} | gzip -1 > ${{READS2TIGS}}
    		{params.racon} -m 8 -x -6 -g -8 -t 4 -q -1 -w 500 ${{READS}} ${{READS2TIGS}} ${{SCAFFOLD}} | gzip -1 > ${{NEWSCAF}}

    		rm ${{READS2TIGS}} &
    		SCAFFOLD=${{NEWSCAF}}
    	done

    	ln -s {output.scaffold} {output.symlink}
		"""




