{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 99,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pysam\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.patches as patches\n",
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 479,
   "metadata": {},
   "outputs": [],
   "source": [
    "RECTANGLE_HEIGHT = 0.2\n",
    "INSERTION_THRESHOLD = 10 # Insertions will not be displayed if length < INSERTION_THRESHOLD\n",
    "INSERTION_THRESHOLD_2 = 100 # Insertions will be highlighted if length >= INSERTION_THRESHOLD_2\n",
    "DELETION_THRESHOLD = 1 # Delections will not be displayed if length < DELETION_THRESHOLD\n",
    "ALIGN_CONTINGENCY = 10 # Blocks will not be displayed if length < ALIGN_CONTINGENCY\n",
    "TARGET_REGION = () # Leave blank for automatic xlim\n",
    "INTER_ASSEMBLY_PADDING = 1\n",
    "X_PADDING = 0.05"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 457,
   "metadata": {},
   "outputs": [],
   "source": [
    "class Canvas:\n",
    "    def __init__(self):\n",
    "        self.assemblies = []\n",
    "        self.y = 0 - INTER_ASSEMBLY_PADDING\n",
    "        \n",
    "    def add_assembly(self, assembly):\n",
    "        assembly.y0 = self.y + INTER_ASSEMBLY_PADDING\n",
    "        self.assemblies.append(assembly)\n",
    "        self.y = max([assembly.y_max for assembly in self.assemblies])\n",
    "        \n",
    "    def draw(self, target_region = None): \n",
    "        # Determine x and y limits\n",
    "        x_min = None\n",
    "        x_max = None\n",
    "        y_max = 0\n",
    "        \n",
    "        for assembly in self.assemblies:\n",
    "            if x_min is None or x_min > assembly.left:\n",
    "                x_min = assembly.left\n",
    "            if x_max is None or x_max < assembly.right:\n",
    "                x_max = assembly.right\n",
    "            if y_max < assembly.y_max + assembly.y0:\n",
    "                y_max = assembly.y_max + assembly.y0\n",
    "        if target_region:\n",
    "            x_min = target_region[0]\n",
    "            x_max = target_region[1]\n",
    "        \n",
    "        # Create figure\n",
    "        plt.figure(figsize = (10, 0.5 * y_max))\n",
    "        \n",
    "        # Plot assemblies\n",
    "        for assembly in self.assemblies:\n",
    "            assembly.draw()\n",
    "        \n",
    "        # Set x and y limits\n",
    "        delta_x = X_PADDING * (x_max - x_min)\n",
    "        plt.xlim(x_min - delta_x, x_max + delta_x)\n",
    "        \n",
    "        plt.ylim(0, y_max + 1)\n",
    "        \n",
    "        # Label assemblies\n",
    "        for assembly in self.assemblies:\n",
    "            plt.text(x_max, assembly.y0 + 0.05, assembly.name, ha = 'right', va = 'bottom')\n",
    "        \n",
    "        # Hide ticks on y axis\n",
    "        plt.tick_params(axis='y', which='both', left=False, right=False, labelleft=False)\n",
    "        \n",
    "        # Misc\n",
    "        plt.xlabel(\"Reference postition\")\n",
    "        plt.tight_layout()\n",
    "                \n",
    "        \n",
    "    def __repr__(self):\n",
    "        output = \"\"\n",
    "        for assembly in self.assemblies:\n",
    "            output += assembly.__repr__()\n",
    "        return output\n",
    "    \n",
    "\n",
    "class Assembly:\n",
    "    def __init__(self, y0 = 0, name = \"\"):\n",
    "        self.rows = []\n",
    "        self.name = name\n",
    "        self.y0 = y0\n",
    "        self.y_max = 0\n",
    "        self.left = None\n",
    "        self.right = None\n",
    "        \n",
    "    def add_contig(self, contig):\n",
    "        if self.left is None or self.left > contig.left:\n",
    "            self.left = contig.left\n",
    "        if self.right is None or self.right < contig.right:\n",
    "            self.right = contig.right\n",
    "            \n",
    "        # Try to insert contig into one of the existing rows\n",
    "        for row in self.rows:\n",
    "            if row.check_overlap(contig) == False:\n",
    "                row.add_contig(contig)\n",
    "                return\n",
    "        \n",
    "        # Otherwise, create a new row and add the contig\n",
    "        y = self.y_max + 1\n",
    "        new_row = self.add_row(y)\n",
    "        self.y_max = y\n",
    "        new_row.add_contig(contig)\n",
    "    \n",
    "\n",
    "            \n",
    "    def add_row(self, y):\n",
    "        new_row = Row(y)\n",
    "        self.rows.append(new_row)\n",
    "        return new_row\n",
    "\n",
    "    def draw(self):\n",
    "        for row in self.rows:\n",
    "            row.draw(self.y0)\n",
    "            \n",
    "        # Draw a horizontal seperating line\n",
    "        plt.gca().axhline(y = self.y0, linewidth=1, color='whitesmoke')\n",
    "\n",
    "    def __repr__(self):\n",
    "        output = \"Assembly {name} (y0 = {y0})\\n\".format(name=self.name, y0=self.y0)\n",
    "        for row in self.rows:\n",
    "            output += row.__repr__()\n",
    "        return output\n",
    "        \n",
    "\n",
    "class Row:\n",
    "    def __init__(self, y):\n",
    "        self.y = y\n",
    "        self.contigs = []\n",
    "        \n",
    "    def check_overlap(self, contig, padding = 10):\n",
    "        overlapped = False\n",
    "        for existing_contig in self.contigs:\n",
    "            if contig.left >= existing_contig.left - padding and contig.left <= existing_contig.right + padding or \\\n",
    "            contig.right >= existing_contig.left - padding and contig.right <= existing_contig.right + padding:\n",
    "                overlapped = True\n",
    "        return overlapped\n",
    "    \n",
    "    def add_contig(self, contig):\n",
    "        self.contigs.append(contig)\n",
    "        \n",
    "    def draw(self, y0):\n",
    "        for contig in self.contigs:\n",
    "            contig.draw(self.y + y0)\n",
    "            \n",
    "    def __repr__(self):\n",
    "        output = \"Row {y}\\n\".format(y = self.y)\n",
    "        for contig in self.contigs:\n",
    "            output += contig.__repr__() + \"\\n\"\n",
    "        return output\n",
    "            \n",
    "            \n",
    "\n",
    "class Contig:\n",
    "    def __init__(self, aligned_pairs):\n",
    "        self.blocks = []\n",
    "        self.left = None\n",
    "        self.right = None\n",
    "        self.width = None\n",
    "        self.insertions = detect_insertions(aligned_pairs)\n",
    "        self.deletions = detect_deletions(aligned_pairs)\n",
    "        \n",
    "    def add_block(self, block):\n",
    "        self.blocks.append(block)\n",
    "        if self.left is None or self.left > block.left:\n",
    "            self.left = block.left\n",
    "        if self.right is None or self.right < block.left:\n",
    "            self.right = block.right\n",
    "        self.width = self.right - self.left\n",
    "            \n",
    "    def draw(self, y):\n",
    "        \n",
    "        # Draw a line across the contig\n",
    "        width = self.right - self.left\n",
    "        line_height = 0.05\n",
    "        ax = plt.gca()\n",
    "        global RECTANGLE_HEIGHT\n",
    "        ax.add_patch(patches.Rectangle((self.left, y + 0.5 * RECTANGLE_HEIGHT - 0.5 * line_height), \n",
    "                                       width, line_height, facecolor = \"firebrick\", edgecolor=None, zorder=0.8))\n",
    "        for block in self.blocks:\n",
    "            block.draw(y)\n",
    "        \n",
    "        if self.insertions:\n",
    "            for insertion in self.insertions:\n",
    "                insertion.draw(y)\n",
    "        \n",
    "        if self.deletions:\n",
    "            for deletion in self.deletions:\n",
    "                deletion.draw(y)\n",
    "        \n",
    "            \n",
    "    def __repr__(self):\n",
    "        return \"Contig {left}-{right} ({n} blocks) \".format(left = self.left, right = self.right, n = len(self.blocks))\n",
    "\n",
    "class Block:\n",
    "    def __init__(self, left, right):\n",
    "        self.left = left\n",
    "        self.right = right\n",
    "        self.width = self.right - self.left\n",
    "        self.y = None\n",
    "        \n",
    "    def draw(self, y):\n",
    "        if self.right - self.left < ALIGN_CONTINGENCY:\n",
    "            return\n",
    "        else:\n",
    "            rectangle_color = \"steelblue\"\n",
    "        ax = plt.gca()\n",
    "        ax.add_patch(patches.Rectangle((self.left, y), self.width, RECTANGLE_HEIGHT, zorder=1, facecolor=rectangle_color, edgecolor=None))\n",
    "        \n",
    "class Insertion:\n",
    "    def __init__(self, position, length):\n",
    "        self.position = position\n",
    "        self.length = length\n",
    "    def draw(self, y):\n",
    "        global INSERTION_THRESHOLD\n",
    "        if self.length < INSERTION_THRESHOLD:\n",
    "            return\n",
    "        if not self.position or not y:\n",
    "            return\n",
    "        if self.length\n",
    "        plt.plot(self.position, y, marker = align_marker(\"^\", valign=\"top\"), color='purple', markersize=10, markeredgewidth=0)\n",
    "        \n",
    "    def __repr__(self):\n",
    "        return \"Insertion(position = {position}, length = {length})\".format(\n",
    "            position = self.position, length = self.length)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 458,
   "metadata": {},
   "outputs": [],
   "source": [
    "class Deletion():\n",
    "    def __init__(self, left, right):\n",
    "        self.left = left\n",
    "        self.right = right\n",
    "        self.width = self.right - self.left\n",
    "    def __repr__(self):\n",
    "        output = \"Deletion(width = {width}) \".format(width = self.width)\n",
    "        return output\n",
    "    def draw(self, y):\n",
    "        if self.width < DELETION_THRESHOLD:\n",
    "            return\n",
    "        ax = plt.gca()\n",
    "        ax.add_patch(patches.Rectangle((self.left, y), self.width, RECTANGLE_HEIGHT, zorder=2, facecolor='red', edgecolor=None))\n",
    "            "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 459,
   "metadata": {},
   "outputs": [],
   "source": [
    "def detect_insertions(aligned_pairs):\n",
    "    # How to distinguish between small deletions and poorly assemblied regions?\n",
    "    length = 0\n",
    "    position = 0\n",
    "    insertions = []\n",
    "    for query, reference in aligned_pairs:\n",
    "        if query and not reference:\n",
    "            if length == 0: # Start of a new insertion              \n",
    "                pass\n",
    "            else: # Continue current insertion\n",
    "                pass\n",
    "            length += 1\n",
    "            \n",
    "        else:\n",
    "            if length == 0: # Matched pair; record position\n",
    "                position = reference\n",
    "            else: # End current insertion\n",
    "                insertions.append(Insertion(position, length))\n",
    "                position = 0\n",
    "                length = 0\n",
    "    return insertions\n",
    "\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 460,
   "metadata": {},
   "outputs": [],
   "source": [
    "def detect_deletions(aligned_pairs):\n",
    "    deletions = []\n",
    "    left = None\n",
    "    right = None\n",
    "    for query, reference in aligned_pairs:\n",
    "        if reference and not query: \n",
    "            if not left: # Start of a new deletion\n",
    "                left = reference\n",
    "                right = left + 1\n",
    "            else: # Continue current deletion\n",
    "                right = reference\n",
    "        else:\n",
    "            if not left: # Matched pair; record position\n",
    "                pass\n",
    "            else: # End current insertion\n",
    "                deletions.append(Deletion(left, right))\n",
    "                left = None\n",
    "                right = None\n",
    "    return deletions\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 461,
   "metadata": {},
   "outputs": [],
   "source": [
    "from matplotlib import markers\n",
    "from matplotlib.path import Path\n",
    "\n",
    "def align_marker(marker, halign='center', valign='middle',):\n",
    "    \"\"\"\n",
    "    create markers with specified alignment.\n",
    "    From: https://stackoverflow.com/questions/26686722/align-matplotlib-scatter-marker-left-and-or-right\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "\n",
    "    marker : a valid marker specification.\n",
    "      See mpl.markers\n",
    "\n",
    "    halign : string, float {'left', 'center', 'right'}\n",
    "      Specifies the horizontal alignment of the marker. *float* values\n",
    "      specify the alignment in units of the markersize/2 (0 is 'center',\n",
    "      -1 is 'right', 1 is 'left').\n",
    "\n",
    "    valign : string, float {'top', 'middle', 'bottom'}\n",
    "      Specifies the vertical alignment of the marker. *float* values\n",
    "      specify the alignment in units of the markersize/2 (0 is 'middle',\n",
    "      -1 is 'top', 1 is 'bottom').\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "\n",
    "    marker_array : numpy.ndarray\n",
    "      A Nx2 array that specifies the marker path relative to the\n",
    "      plot target point at (0, 0).\n",
    "\n",
    "    Notes\n",
    "    -----\n",
    "    The mark_array can be passed directly to ax.plot and ax.scatter, e.g.::\n",
    "\n",
    "        ax.plot(1, 1, marker=align_marker('>', 'left'))\n",
    "\n",
    "    \"\"\"\n",
    "\n",
    "    if isinstance(halign, (bytes, str)):\n",
    "        halign = {'right': -1.,\n",
    "                  'middle': 0.,\n",
    "                  'center': 0.,\n",
    "                  'left': 1.,\n",
    "                  }[halign]\n",
    "\n",
    "    if isinstance(valign, (bytes, str)):\n",
    "        valign = {'top': -1.,\n",
    "                  'middle': 0.,\n",
    "                  'center': 0.,\n",
    "                  'bottom': 1.,\n",
    "                  }[valign]\n",
    "\n",
    "    # Define the base marker\n",
    "    bm = markers.MarkerStyle(marker)\n",
    "\n",
    "    # Get the marker path and apply the marker transform to get the\n",
    "    # actual marker vertices (they should all be in a unit-square\n",
    "    # centered at (0, 0))\n",
    "    m_arr = bm.get_path().transformed(bm.get_transform()).vertices\n",
    "\n",
    "    # Shift the marker vertices for the specified alignment.\n",
    "    m_arr[:, 0] += halign / 2\n",
    "    m_arr[:, 1] += valign / 2\n",
    "\n",
    "    return Path(m_arr, bm.get_path().codes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 462,
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_bam_file(bam_file, assembly_name = None):\n",
    "    if not assembly_name: \n",
    "        # Use bamfile name without extension as assembly name\n",
    "        base_name = os.path.basename(bam_file)\n",
    "        assembly_name = os.path.splitext(base_name)[0]\n",
    "    assembly = Assembly(name = assembly_name)\n",
    "    bam = pysam.AlignmentFile(bam_file, 'rb')\n",
    "    \n",
    "    contigs = []\n",
    "    for contig_data in bam.fetch():\n",
    "        aligned_pairs = contig_data.get_aligned_pairs()\n",
    "        contig = Contig(aligned_pairs)\n",
    "        blocks_data = contig_data.get_blocks()\n",
    "        for block_data in blocks_data:\n",
    "            left, right = block_data[0], block_data[1]\n",
    "            block = Block(left, right)\n",
    "            contig.add_block(block)     \n",
    "        contigs.append(contig)\n",
    "    \n",
    "    # Sort contigs by width to ensure longer contigs are plotted before shorter contigs\n",
    "    contigs.sort(key = lambda contig: contig.width, reverse=True)\n",
    "    for contig in contigs:\n",
    "        assembly.add_contig(contig)\n",
    "    return assembly"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 478,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAsgAAAEYCAYAAABBfQDEAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4xLjEsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy8QZhcZAAAgAElEQVR4nO3de5SdVZnn8d+TFKZASFcgwkAw17YByXCREnC4VLwsgh0bmQEG0ggRZBxQ0t2u5UzT7QQysljTIpBa0O1AI6QAsWlhmuISMpGRGEFytCuYgChtG6gK0QgKnKpK46ncnvnjvKfYdTj323vqnO9nraycs99373fv593n1FP7vHVec3cBAAAASJsSdwcAAACAZkKCDAAAAARIkAEAAIAACTIAAAAQIEEGAAAAAh2FNs6cOdPnzp3boK4AAAAAjbNp06bfufv7sssLJshz587VwMBA/XoFAAAAxMTMhnKVc4kFAAAAECBBBgAAAAIkyAAAAECABBkAAAAIkCADAAAAARJkAAAAIECCDAAAAARIkAEAAIAACTIAAAAQIEEGAAAAAiTIAAAAQIAEGQAAAAiQIAMAAAABEmQAAAAgQIIMAAAABEiQAQAAgAAJMgAAABAgQQYAAAACJMgAAABAoCPuDgAAgMZafP2ad5WtW7Ekhp4AzYkVZAAAACBAggwAAAAESJABAACAAAkyAAAAECBBBgAAAAJ8iwUAAG2Gb6wACmMFGQAAAAiQIAMAAAABEmQAAAAgQIIMAAAABEiQAQAAgAAJMgAAABAgQQYAAAACJMgAAABAgAQZAAAACJAgAwAAAAESZAAAACBAggwAAAAESJABAACAAAkyAAAAECBBBgAAAAIkyAAAAECABBkAAAAIkCADAAAAARJkAAAAINARdwcAAEC8Fl+/Jmf5uhVLGtwToDmwggwAAAAESJABAACAAAkyAAAAECBBBgAAAAIkyAAAAECAb7EAAKDNrVuxRMnBpEa2j0iSph85PeYeAfEiQQYAoM2lkindffrdGv31qGTSgYcdqCkdU3TV81eps6sz7u4BDcclFgAAtLlEb0KjvxqVXNI+aeeOnRp5dUSJ3kTcXQNiQYIMAEAbSyVTevaWZ3NuS6xKKJVMNbhHQPxIkAEAaGOJ3oR2j+7OuW1sZIxVZLQlEmQAANpYZ1enDjnqEHXs3yGbapJJ1mHa/5D9tWDxAq5BRlsyd8+7sbu72wcGBhrYHQAAAKAxzGyTu3dnl7OCDAAAAARIkAEAAIAACTIAAAAQIEEGAAAAAiTIAAAAQIAEGQAAAAiQIAMAAAABEmQAAAAgQIIMAAAABEiQAQAAgAAJMgAAABAgQQYAAAACJMgAAABAgAQZAAAACHTE3QGgnhZfvyZn+boVS2redi3abHfNFtNm6w9QD/neJzMaMe8L9YHXHeLACjIAAAAQIEEGAAAAAiTIAAAAQIAEGQAAAAiYu+fd2N3d7QMDAw3sDlB/ycGkuuZ21bSdbc9s0+zTZ1fdJibGslbnqlb9AVpdcjCpzfdu1vyPzWfeoy2Y2SZ3784uZwUZbSWVTKlvUZ9SyVTN2kkOJdW3qE/JoWSNetm+wljW6lzVqj9Aq0slU7rz5Du14boN6uth3qO9kSCjrSR6ExoeGlaiN1GzdvqX9cv3uvqX9deol+0rjGWtzlWt+gO0ukRvQm//9m1Jku9j3qO9kSCjbaSSqfFkK9GbqHhlMmzn2Vue1dCGIUnS0IYhVlyqkBxKTojlxls2SqruXNWyP5xbtLJUMqWnv/70hDLmPdoZCTLaRqI3obHhMUnS2PBYxSuTYTu7R3dP2MaKS+WyY7drdJek6s5VLfvDuUUrS/QmtO/tfe8qZ96jXXEnPbSNzq5O9VzXM+F5te389IGfat+efZp+5PT0thmVtYl07Ob0zJEkjWwf0ZSOKVp40cL0tgrPVa36k3kOtKrOrk5Nec8U7dsVJMlTmPdoX3yLBQAAANoS32IBAAAAlIAEGQAAAAiQIAMAAAABEmQAAAAgQIIMAAAABEiQAQAAgAAJMgAAABAgQQYAAAACJMgAAABAgAQZAAAACJAgAwAAAAESZAAAACBAggwAAAAESJABAACAQEfcHQBqYfH1a0rab92KJRW3nV03+5iVtN1q8p2HUmKTq25YL9xeTqyLtVvreij99TjZTbb5UO15qfV4S+nPZIsxWgcryAAAAECABBkAAAAIkCADAAAAARJkAAAAIGDunndjd3e3DwwMNLA7AIDJLjmYVNfcrqJlta5bL3EfPw7bntmm2afPbsix8sW3HeOOxjOzTe7enV3OCjIAoGZSyZT6FvUplUwVLKt13XqJ+/hxSA4l1beoT8mhZN2PlS++7Rh3NBcSZABAzSR6ExoeGlaiN1GwrNZ16yXu48ehf1m/fK+rf1l/3Y+VL77tGHc0FxJkAEBNpJKp8YQm0ZtQKpnKWVbruvUS9/HjkBxKamjDkCRpaMNQXVeR88W3HeOO5kOCDACoiURvQmPDY5KkseExJXoTOctqXbde4j5+HLJXjeu5ipwvvu0YdzQf7qQHAKiJzq5O9VzXM+G5pJxltaxbL/n61Mo6Z3RqTs+cCc/rdqw88W3HuKP58C0WAAAAaEt8iwUAAABQAhJkAAAAIECCDAAAAARIkAEAAIAACTIAAAAQIEEGAAAAAiTIAAAAQIAEGQAAAAiQIAMAAAABEmQAAAAgQIIMAAAABEiQAQAAgAAJMgAAABAgQQYAAAACHXF3INvi69fE3QVMcutWLMlZXu+5le+4jexDsyklJqE4zlG7nZNGKffcT0aTee6sW7Gk4v5XU7dQm1LhmBaaU5P5XOAdzfS+wQoyAAAAECBBBgAAJXtswYK4uwDUHQkyAAAo2Z9s3Rp3F4C6I0EGAAAAAubueTd2d3f7wMBAA7tTe8nBpLrmdo3/X+r+ldr2zDbNPn12xfWbXbXxKVQ/OZiUpKrar6Vqx4rqFZsv7Xh+Gjnuco/VbuckHG++sZdant2WVP57YSXxb4Zz1gx9aGbVxKdY3XJzpFZkZpvcvTu7vKVXkFPJlPoW9Sk5lFTfoj6lkqmS9i+2Xz6Z4ySHkhXVb3bVxqdQ/VQypdVnrtbqM1dX3H4tVTtWVK/YfGnH89PIcZd7rHY7J+F484291PLstip5L6wk/s1wzpqhD82smvgUq1tujtRuWjpBTvQmNDw0rP5l/RoeGlaiN1HS/sX2y6d/Wb98r6t/WX9F9ZtdtfEpVD/Rm9DIqyMaeXWk4vZrqdqxonrF5ks7np9GjrvcY7XbOQnHm2/spZZnt1XJe2El8W+Gc9YMfWhm1cSnWN1yc6R207IJciqZGj/ZQxuGJKUnQ6HfpMI3q3J/k0oOJcePM7RhqOVWkauNT6H6qWRKG1dtHH+eWFV++7VU7VhRvWLzpR3PTyPHXe6x2u2chOPduGqjEqvePfZ8MckuTw4lJ7S18Zby3wsriX8znLNm6EMzqyY+xeqWmyO1o5ZNkBO9CY0Nj00oGxseK/ibVGb/Qvvlk71q3GqryNXGp1D9RG9Cu0Z2jT8fGym//VqqdqyoXrH50o7np5HjLvdY7XZOwvHuGtmlsZF3jz1fTLLL+5f1T2hr12j574WVxL8Zzlkz9KGZVROfYnXLzZHaUdPdSa9WOrs61XNdj7YntmtPao86Ojt05KlHqrOrs+D+4fOyjjejU3N65kx43kqqjk+B+p1dnVqweIH2pPZIkjo6O8puv5aqHSuqV2y+tOP5aeS4yz1Wu52TcLzbE9slSUeeeuT4tux9CpX/ZstvJrSVHEzqwH93oKTS3wsriX8znLNm6EMzqyY+xeqWmyO1o5b/FgsAAAAgl7b8FgsAAACgXCTIAAAAQIAEGQAAAAiQIAMAAAABEmQAAAAgQIIMAAAABEiQAQAAgAAJMgAAABAgQQYAAAACJMgAAABAgAQZAAAACJAgAwAAIK97771Xxx13nI4//nhdcskleuyxx3TKKafoxBNP1Cc+8Qm99tprkqSVK1fq8ssv16JFizR//nzdeuutkqTBwUEtXLhwvL2bbrpJK1eujGMoJeuIuwMAAABoTi+++KJuuOEG/fCHP9TMmTP15ptvysyUSCRkZvrmN7+pG2+8UTfffLMk6aWXXtL69es1Ojqqo446SldddVXMI6gMCTIAAAByeuqpp3T++edr5syZkqSDDz5YL7zwgi688ELt2LFDu3bt0rx588b3X7JkiaZNm6Zp06bp0EMPHV9dnmy4xAIAAAA5ubvMbELZ8uXLdfXVV+uFF17QHXfcoVQqNb5t2rRp44+nTp2qPXv2qKOjQ/v27RsvD/dvViTIAAAAyOnjH/+4vvOd7+iNN96QJL355psaHh7WrFmzJEn33HNP0TYOO+wwvf7663rjjTc0Njamxx9/vK59rgUusQAAAEBOxx57rL7yla+op6dHU6dO1YknnqiVK1fqggsu0KxZs3TqqafqlVdeKdjGfvvtp2uvvVannHKK5s2bp6OPPrpBva+cuXvejd3d3T4wMNDA7gAAAACNYWab3L07u5xLLAAAAIBAwUss9u3bp7fffrtRfQEAAABiVzBBnjJlig444IBG9QUAAACIHZdYAAAAAAESZAAAACBAggwAAAAESJABAACAAAkyAAAAECBBBgAAAAIkyAAAAECABBkAAAAIkCADAAAAARJkAAAAIECCDAAAAARIkAEAAIAACTIAAAAQIEEGAAAAAiTIAAAAQIAEGQAAAAiQIAMAAAABEmQAAAAgQIIMAAAABDri7gAmp8XXr6m6jXXXfkqLv/p4DXqTp/0VS+rWtvTuGNT7eG3BbMKcCGNaypzLNafiOC+1eH2UinkHALXHCjIAAAAQIEEGAAAAAiTIAAAAQIAEGQAAAAiYu+fd2N3d7QMDAw3sDlpdcjCprrldebdJyru92RQaC2qjkhi3+nlp9fEBQCOZ2SZ3784uZwUZDZNKptS3qE+pZCrnttVnrtbqM1fn3N5sCo0FtVFJjFv9vLT6+ACgWZAgo2ESvQkNDw0r0ZvIuW3k1RGNvDqSc3uzKTQW1EYlMW7189Lq4wOAZkGCjIZIJVPjP9QTvYkJK2CpZEobV20cf55YlWjqFbJCY0FtVBLjVj8vrT4+AGgmJMhoiERvQmPDY5KkseGxCStgid6Edo3sGn8+NjLW1CtkhcaC2qgkxq1+Xlp9fADQTLiTHhqis6tTPdf1THgePl6weIH2pPZIkjo6OyZsbzaFxoLaqCTGrX5eWn18ANBM+BYLAAAAtCW+xQIAAAAoAQkyAAAAECBBBgAAAAIkyAAAAECABBkAAAAIkCADAAAAARJkAAAAIECCDAAAAARIkAEAAIAACTIAAAAQIEEGAAAAAiTIAAAAQIAEGQAAAAiQIAMAAACBjrg7kG3x9WtiO/a6FUtiPf5ksm7Fkri7UJVKzvO6az+lxV99vOJjZuZXte20skKxiWPOZc+TTB9KnT+N6HMzv2flGn8z9zefUs7jZBxXRvb4io2l1j8ryz0+KjNZfvY0S37BCjIAAAAQIEEGAAAle2zBgri7ANQdCTIAAAAQIEEGAAAl+5OtW+PuAlB3JMgAAABAwNw978bu7m4fGBhoYHdqLzmYVNfcroYdb9sz2zT79NkNO16jNTqeaD3JwaQklTSPmG/xKjf+9Tpfk2Ee5OtjueWZbVJpr5F62PbMNk0/cnrDjv/jv/uxTv7iyQ051mRUzfwvte5keI3Vi5ltcvfu7PKWXkFOJVPqW9SnVDLVkOMlh5LqW9Sn5FCyIcdrtEbHE60nlUxp9ZmrtfrM1UXnEfMtXuXGv17nazLMg3x9LLc8s63U10g9JIeSWt2zWneffndDjr/tmW1ae/VabXtmW92PNRlVM/9LrTsZXmNxaOkEOdGb0PDQsBK9iYYcr39Zv3yvq39Zf0OO12iNjidaT6I3oZFXRzTy6kjRecR8i1e58a/X+ZoM8yBfH8stz2wr9TVSD/3L+qV90uivRhty/G9/6tsT/sdE1cz/UutOhtdYHFo2QU4lU+MnO9GbqPtvRsmhpIY2DEmShjYMtdwqcqPjidaTSqa0cdXG8eeJVfnnEfMtXuXGv17nazLMg3x9LLc8s63U10g9hD/HJGnjLRvrevxtz2zT2PCYJGlseIxV5CzVzP9S606G11hcWjZBTvQmJrzw6v2bUfaqcautIjc6nmg9id6Edo3sGn8+NpJ/HjHf4lVu/Ot1vibDPMjXx3LLM9tKfY3UQ/bPrV2ju+p6/OxVY1aRJ6pm/pdadzK8xuLSdLearpXOrk71XNcz4XldjzejU3N65kx43koaHU+0ns6uTi1YvEB7UnskSR2dHXnnEfMtXuXGv17nazLMg3x9LLc887jU10g9dM7o1IwFM7Rvz76GHL9j/46Cz9tdNfO/1LqT4TUWl5b/FgsAAAAgl7b8FgsAAACgXCTIAAAAQIAEGQAAAAiQIAMAAAABEmQAAAAgQIIMAAAABEiQAQAAgAAJMgAAABAgQQYAAEDZbr31Vh1zzDGaNWuWrr766ri7U1Pc1xEAAABl+8Y3vqG1a9dqw4YNarU7L7OCDAAAgLJceeWVevnll3XOOeforbfekiSNjo5q3rx52r17tyRpZGREc+fO1e7du7V161adffbZOumkk3TGGWfopZdeirP7RZEgAwAAoCy33367jjjiCK1fv14zZsyQJB100EFatGiR1qxZI0l64IEHdN5552m//fbT5z//ed12223atGmTbrrpJn3hC1+Is/tFcYkFAAAAauKKK67QjTfeqHPPPVerV6/WnXfeqZ07d+rZZ5/VBRdcML7f2NhYjL0sjgQZAAAANXHaaadpcHBQGzZs0N69e7Vw4UKNjIyoq6tLmzdvjrt7JeMSCwAAANTMpZdeqqVLl+qyyy6TJE2fPl3z5s3Tgw8+KElyd23ZsiXOLhZFggwAAICaufjii/XWW29p6dKl42X333+/7rrrLh1//PE69thj9cgjj8TYw+LM3fNu7O7u9lb72g4AAADUz0MPPaRHHnlE9913X9xdKcrMNrl7d3Y51yADAACgJpYvX661a9fqiSeeiLsrVSFBBgAAQE3cdtttcXehJgpeYmFmv5U01LjuNLWZkn4XdycmMeJXOWJXOWJXHeJXOWJXOWJXHeJXnjnu/r7swoIJMt5hZgO5rlFBaYhf5Yhd5YhddYhf5Yhd5YhddYhfbfAtFgAAAECABBkAAAAIkCCX7u/j7sAkR/wqR+wqR+yqQ/wqR+wqR+yqQ/xqgGuQAQAAgAAryAAAAECABBkAAAAItGWCbGbvN7P1ZvZzM3vRzP48xz5/YGaPmdmWaJ/Lgm3LzOxfo3/LgvKTzOwFM/ulmd1qZtaoMTVKNbEzsxPMbGNU9ryZXRjU6TOzV8xsc/TvhEaOq1FqMPf2BjF6NCifZ2Y/iubkP5rZexo1pkapcu59NIjbZjNLmdm50baWn3slxm6GmT0cvTZ/bGYLg21nm9m/RO9t1wTlzDvlj12huma20sx+Fcy7P27kuBqlBnNvMPq5utnMBoLyg83syWjuPWlmMxo1pkapcu4dlfWeN2JmfxFta4u5VzV3b7t/kg6X9KHo8UGSfiHpg1n7/LWkr0WP3yfpTUnvkXSwpJej/2dEj2dE+/1Y0kckmaS1kj4Z91ibLHZ/JOkDUfkRknZI6oqe90k6P+7xNXP8ouc787T7HUkXRY9vl3RV3GNtttgF+xwclR8QPW/5uVdi7L4u6bro8dGSvhc9nippq6T50et4S6Yu865o7PLWlbRS0pfjHl8zxy96PihpZo52b5R0TfT4mszrvpX+VRu7YJ+pkn6j9A0x2mbuVfuvLVeQ3X2Huz8XPR6V9HNJs7J3k3SQmZmkA5X+gbpH0mJJT7r7m+7+lqQnJZ1tZodLmu7uGz09A++VdG5jRtQ41cTO3X/h7v8a1f21pNeVTmLaRpVzL6dov49JeigqukfMvUKxO1/SWnd/u85dbholxu6Dkr4X7fOSpLlmdpikkyX90t1fdvddkh6Q9Gnm3QQ5Y1di3ZZW5dwr5NNKzzmJuVcsdh+XtNXduTNyGdoyQQ6Z2VxJJ0r6Udamv5V0jKRfS3pB0p+7+z6lJ+erwX7bo7JZ0ePs8pZVQezCuicrvRq1NSi+IfqYaJWZTatXv5tFhfHrNLMBM0tkLhGQdIikpLtnEkHmXoG5J+kiSf+QVdY2c69A7LZI+k/RPidLmiPpSOV/z2PevSNf7IrVvTqad3e34iUC2SqMn0v6rpltMrPPB3UOc/cdUjqRlHRo/Xoev2rmnnK/57XV3KtEWyfIZnagpP8j6S/cfSRr82JJm5W+FOAESX9rZtOVvnwimxcob0kVxi5T93BJ90m6LEhe/krpj4c+rPRH4H9Z3xHEq4r4zfb0LUT/VFKvmS0Qcy9Uytz795LWBXXaZu4Vid3fSJphZpslLZf0E6VX33nPU8WxK1T3f0taoPQ83SHp5vqOIF5VxO80d/+QpE9K+qKZndmoPjeLKufeeySdI+nBoE5bzb1KtW2CbGb7KT3h7nf3f8qxy2WS/snTfinpFaV/iG6X9P5gvyOVXq3arom/tWXKW04VsVOUrKyR9D/cPZGpEH2U5O4+Jmm10h/rtqRq4hddmiJ3f1nS95VeUfidpC4z64jqM/dyxC7ynyU97O67MwXtMveKxc7dR9z9Mnc/QdKlSl/+9Iryv+cx7yIFYpe3rru/5u57o0WCO9Wi806qLn7Be97rkh7WO3F6LfqFN/OL7+t1H0gMqold5JOSnnP314I6bTP3qtGWCXJ07dxdkn7u7rfk2W2b0tftKLqe5yil/yBvnaSzor8cnSHpLEnroo94Rs3s1Kj9SyU9UuehNFw1sYt+k31Y0r3uHv42q+CNzpS+luyn9RlBvKqM34zMx/9mNlPSaZJ+Fl3zvl7pa2slaZmYe9mv24ylyvqosR3mXimxM7Mue+dbKK6Q9INoteqfJX3A0t9Y8R6lP659lHk3YZ+csStUNzPvIv9RLTjvpKrj914zOyja571K/7zNxOlRpeecxNzL9brNyPueF2nZuVettryTnpmdLulppa9RzHzE/9eSZkuSu99uZkco/dfthyv9UeLfuPu3ovqXR/tL0g3uvjoq747q7K/0t1gs9xYLcDWxM7PPKL1C92LQ5GfdfbOZPaX0b76m9EfkV7r7zgYMqaGqjN9/kHRHVG+KpF53vytqd77Sfzx1sNIfsX0mWhFtGTV43c6V9ENJ7w+vS26HuVdi7D6i9B8X75X0M0mf8/QfIsvSXwPVq/Rfw9/t7jdE5cy7ArHLV9fdnzCz+5T+iNuV/qaG/5q5praVVBm/+UovqkhSh6RvB3PvEKW/RWW20r8YX+DubzZmVI1Rg9ftAUr//cB8dx8O2m2LuVettkyQAQAAgHza8hILAAAAIB8SZAAAACBAggwAAAAESJABAACAAAkyAAAAJpXoLoCvm1nRr6mz9F1SN0f/fmFmyWJ1SJABtBQz2xu9Cf7UzB4zs64S6vyZmf3czO5vRB/jYmaLoq8LzDy/0swujR5/NvqavMy2b5rZB+PoJwCUoE/S2aXs6O5fcvcTohuq3CYp142mJiBBBtBqfh+9ES6U9KakL5ZQ5wuS/tjdLy7lAMHd4yabRZLGE2R3v93d742eflbpW3Rntl3h7j9raO8AoETu/gOl3+PHmdkCM/u/ZrbJzJ42s6NzVH3XzVNyIUEG0Mo2SpqVeWJm/83M/tnMnjez/xmV3S5pvqRHzexL0d277o72+4mZfTra77Nm9qCZPSbpuwXamxutRt9pZi+a2XfNbP9o2x+a2f8zsy1m9pyZLcjXTjYz22lmN0f1vmdm74vKTzCzRFT3YUvf4TOzKv6zqPyB6EYpV0r6UrTCfoaZrTSzL5vZ+ZK6Jd0fbdvfzL4f3fxIZrbUzF6IVuW/ltWnG6LxJCx990IAiMvfK32TtpMkfVnSN8KNZjZH0jxJTxVriAQZQEsys6lK33b60ej5WZI+IOlkpe8idZKZnenuV0r6taSPuvsqSV+R9JS7f1jSRyV93dK3uZWkj0ha5u4fy9detN8HJP2dux8rKSnpvKj8/qj8eKVXcncUaSf0XknPufuHJG2QdF1Ufq+kv3T345S+41am/BpJJ0blV7r7oKTbJa2KVtifzjTs7g9JGpB0cbTt90Ecj5D0NUkfi/r3YTM7N+hTIhrPDyT9l9xnAwDqy8wOVPp99UEz26z0nWcPz9rtIkkPufveYu2RIANoNftHb45vKH0L5Cej8rOifz+R9Jyko5VOTLOdJemaqI3vS+pUdGtXSU8Gt7Mt1N4r7r45erxJ0lwzO0jSLHd/WJLcPeXub5fRr32S/jF6/C1Jp5vZH0jqcvcNUfk9kjLJ9fNKrwh/RtKe3KEqyYclfd/df+vue5RO8jPH2CXp8XCcVRwHAKoxRVIyc61x9O+YrH0uUgmXV0jpe5sDQCv5vbufECWPjyt9DfKtkkzS/3L3O4rUN0nnufu/TCg0O0XSv2Xt9672oksZxoKivZL2j/bPd7xS+pXNi2xfonQie46kFWZ2bJntZ+TrtyTtdvdMP/aKnykAYuLuI2b2ipld4O4PmplJOs7dt0iSmR0laYbSl94VxQoygJbk7sOS/kzSl81sP0nrJF0efQwnM5tlZofmqLpO0vLozVVmdmKeQ5TaXqY/I5K2Zy5PMLNpZnZAGe1MkXR+9PhPJT0TjfEtMzsjKr9E0gYzmyLp/e6+XtJ/l9Ql6UBJo5IOytPFfNt+JKnHzGZGl60sVfoSDwCIjZn9g9LJ7lFmtt3MPifpYkmfM7Mtkl6U9OmgylJJDwS/1BfEb/sAWpa7/yR6o7zI3e8zs2MkbYxy352SPiPp9axq10vqlfR8lCQPSvpUjra/m6e9Qte2XSLpDjP7qqTdki4o0E52v/5N0rFmtknSsKQLo/Jlkm6Pku2XJV0maaqkb0Wr6Kb0dcfJ6A8MH4r+8HB5Vvt9UTu/V/pa68w4d5jZX0laH7X1hLs/UmCMAFB37r40z6acX/3m7ivLad9KTKQBADEys53ufmDc/QCAdsAlFgAAAECAFWQAAAAgwJyfndEAAAAnSURBVAoyAAAAECBBBgAAAAIkyAAAAECABBkAAAAIkCADAAAAgf8P8OOtznvvNNgAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 720x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "assembly = read_bam_file(\"data/IL27/flye.bam\")\n",
    "assembly2 = read_bam_file(\"data/IL27/canu.bam\")\n",
    "canvas = Canvas()\n",
    "canvas.add_assembly(assembly)\n",
    "canvas.add_assembly(assembly2)\n",
    "canvas.draw(target_region = (27888679, 29888679))\n",
    "\n",
    "plt.savefig(\"1.svg\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 472,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Help on AlignedSegment object:\n",
      "\n",
      "class AlignedSegment(builtins.object)\n",
      " |  AlignedSegment(AlignmentHeader header=None)\n",
      " |  Class representing an aligned segment.\n",
      " |  \n",
      " |      This class stores a handle to the samtools C-structure representing\n",
      " |      an aligned read. Member read access is forwarded to the C-structure\n",
      " |      and converted into python objects. This implementation should be fast,\n",
      " |      as only the data needed is converted.\n",
      " |  \n",
      " |      For write access, the C-structure is updated in-place. This is\n",
      " |      not the most efficient way to build BAM entries, as the variable\n",
      " |      length data is concatenated and thus needs to be resized if\n",
      " |      a field is updated. Furthermore, the BAM entry might be\n",
      " |      in an inconsistent state.\n",
      " |  \n",
      " |      One issue to look out for is that the sequence should always\n",
      " |      be set *before* the quality scores. Setting the sequence will\n",
      " |      also erase any quality scores that were set previously.\n",
      " |  \n",
      " |      Parameters\n",
      " |      ----------\n",
      " |  \n",
      " |      header -- :class:`~pysam.AlignmentHeader` object to map numerical\n",
      " |                identifiers to chromosome names. If not given, an empty\n",
      " |                header is created.\n",
      " |  \n",
      " |  Methods defined here:\n",
      " |  \n",
      " |  __copy__(...)\n",
      " |      AlignedSegment.__copy__(self)\n",
      " |  \n",
      " |  __deepcopy__(...)\n",
      " |      AlignedSegment.__deepcopy__(self, memo)\n",
      " |  \n",
      " |  __eq__(self, value, /)\n",
      " |      Return self==value.\n",
      " |  \n",
      " |  __ge__(self, value, /)\n",
      " |      Return self>=value.\n",
      " |  \n",
      " |  __gt__(self, value, /)\n",
      " |      Return self>value.\n",
      " |  \n",
      " |  __hash__(self, /)\n",
      " |      Return hash(self).\n",
      " |  \n",
      " |  __init__(self, /, *args, **kwargs)\n",
      " |      Initialize self.  See help(type(self)) for accurate signature.\n",
      " |  \n",
      " |  __le__(self, value, /)\n",
      " |      Return self<=value.\n",
      " |  \n",
      " |  __lt__(self, value, /)\n",
      " |      Return self<value.\n",
      " |  \n",
      " |  __ne__(self, value, /)\n",
      " |      Return self!=value.\n",
      " |  \n",
      " |  __reduce__ = __reduce_cython__(...)\n",
      " |      AlignedSegment.__reduce_cython__(self)\n",
      " |  \n",
      " |  __setstate__ = __setstate_cython__(...)\n",
      " |      AlignedSegment.__setstate_cython__(self, __pyx_state)\n",
      " |  \n",
      " |  __str__(...)\n",
      " |      return string representation of alignment.\n",
      " |      \n",
      " |      The representation is an approximate :term:`SAM` format, because\n",
      " |      an aligned read might not be associated with a :term:`AlignmentFile`.\n",
      " |      As a result :term:`tid` is shown instead of the reference name.\n",
      " |      Similarly, the tags field is returned in its parsed state.\n",
      " |      \n",
      " |      To get a valid SAM record, use :meth:`to_string`.\n",
      " |  \n",
      " |  compare(...)\n",
      " |      AlignedSegment.compare(self, AlignedSegment other)\n",
      " |      return -1,0,1, if contents in this are binary\n",
      " |              <,=,> to *other*\n",
      " |  \n",
      " |  get_aligned_pairs(...)\n",
      " |      AlignedSegment.get_aligned_pairs(self, matches_only=False, with_seq=False)\n",
      " |      a list of aligned read (query) and reference positions.\n",
      " |      \n",
      " |              For inserts, deletions, skipping either query or reference\n",
      " |              position may be None.\n",
      " |      \n",
      " |              Padding is currently not supported and leads to an exception.\n",
      " |      \n",
      " |              Parameters\n",
      " |              ----------\n",
      " |      \n",
      " |              matches_only : bool\n",
      " |                If True, only matched bases are returned - no None on either\n",
      " |                side.\n",
      " |              with_seq : bool\n",
      " |                If True, return a third element in the tuple containing the\n",
      " |                reference sequence. Substitutions are lower-case. This option\n",
      " |                requires an MD tag to be present.\n",
      " |      \n",
      " |              Returns\n",
      " |              -------\n",
      " |      \n",
      " |              aligned_pairs : list of tuples\n",
      " |  \n",
      " |  get_blocks(...)\n",
      " |      AlignedSegment.get_blocks(self)\n",
      " |      a list of start and end positions of\n",
      " |             aligned gapless blocks.\n",
      " |      \n",
      " |             The start and end positions are in genomic\n",
      " |             coordinates.\n",
      " |      \n",
      " |             Blocks are not normalized, i.e. two blocks\n",
      " |             might be directly adjacent. This happens if\n",
      " |             the two blocks are separated by an insertion\n",
      " |             in the read.\n",
      " |  \n",
      " |  get_cigar_stats(...)\n",
      " |      AlignedSegment.get_cigar_stats(self)\n",
      " |      summary of operations in cigar string.\n",
      " |      \n",
      " |              The output order in the array is \"MIDNSHP=X\" followed by a\n",
      " |              field for the NM tag. If the NM tag is not present, this\n",
      " |              field will always be 0.\n",
      " |      \n",
      " |              +-----+--------------+-----+\n",
      " |              |M    |BAM_CMATCH    |0    |\n",
      " |              +-----+--------------+-----+\n",
      " |              |I    |BAM_CINS      |1    |\n",
      " |              +-----+--------------+-----+\n",
      " |              |D    |BAM_CDEL      |2    |\n",
      " |              +-----+--------------+-----+\n",
      " |              |N    |BAM_CREF_SKIP |3    |\n",
      " |              +-----+--------------+-----+\n",
      " |              |S    |BAM_CSOFT_CLIP|4    |\n",
      " |              +-----+--------------+-----+\n",
      " |              |H    |BAM_CHARD_CLIP|5    |\n",
      " |              +-----+--------------+-----+\n",
      " |              |P    |BAM_CPAD      |6    |\n",
      " |              +-----+--------------+-----+\n",
      " |              |=    |BAM_CEQUAL    |7    |\n",
      " |              +-----+--------------+-----+\n",
      " |              |X    |BAM_CDIFF     |8    |\n",
      " |              +-----+--------------+-----+\n",
      " |              |B    |BAM_CBACK     |9    |\n",
      " |              +-----+--------------+-----+\n",
      " |              |NM   |NM tag        |10   |\n",
      " |              +-----+--------------+-----+\n",
      " |      \n",
      " |              If no cigar string is present, empty arrays will be returned.\n",
      " |      \n",
      " |              Parameters\n",
      " |              ----------\n",
      " |      \n",
      " |              Returns\n",
      " |              -------\n",
      " |      \n",
      " |              arrays : two arrays. The first contains the nucleotide counts within\n",
      " |                 each cigar operation, the second contains the number of blocks for\n",
      " |                 each cigar operation.\n",
      " |  \n",
      " |  get_forward_qualities(...)\n",
      " |      AlignedSegment.get_forward_qualities(self)\n",
      " |      return base qualities of the read sequence.\n",
      " |              \n",
      " |              Reads mapping to the reverse strand will be reversed.\n",
      " |  \n",
      " |  get_forward_sequence(...)\n",
      " |      AlignedSegment.get_forward_sequence(self)\n",
      " |      return the original read sequence.\n",
      " |              \n",
      " |              Reads mapping to the reverse strand will be reverse\n",
      " |              complemented.\n",
      " |      \n",
      " |              Returns None if the record has no query sequence.\n",
      " |  \n",
      " |  get_overlap(...)\n",
      " |      AlignedSegment.get_overlap(self, uint32_t start, uint32_t end)\n",
      " |      return number of aligned bases of read overlapping the interval\n",
      " |              *start* and *end* on the reference sequence.\n",
      " |      \n",
      " |              Return None if cigar alignment is not available.\n",
      " |  \n",
      " |  get_reference_positions(...)\n",
      " |      AlignedSegment.get_reference_positions(self, full_length=False)\n",
      " |      a list of reference positions that this read aligns to.\n",
      " |      \n",
      " |              By default, this method only returns positions in the\n",
      " |              reference that are within the alignment. If *full_length* is\n",
      " |              set, None values will be included for any soft-clipped or\n",
      " |              unaligned positions within the read. The returned list will\n",
      " |              thus be of the same length as the read.\n",
      " |  \n",
      " |  get_reference_sequence(...)\n",
      " |      AlignedSegment.get_reference_sequence(self)\n",
      " |      return the reference sequence in the region that is covered by the\n",
      " |              alignment of the read to the reference.\n",
      " |      \n",
      " |              This method requires the MD tag to be set.\n",
      " |  \n",
      " |  get_tag(...)\n",
      " |      AlignedSegment.get_tag(self, tag, with_value_type=False)\n",
      " |      \n",
      " |      retrieves data from the optional alignment section\n",
      " |      given a two-letter *tag* denoting the field.\n",
      " |      \n",
      " |      The returned value is cast into an appropriate python type.\n",
      " |      \n",
      " |      This method is the fastest way to access the optional\n",
      " |      alignment section if only few tags need to be retrieved.\n",
      " |      \n",
      " |      Possible value types are \"AcCsSiIfZHB\" (see BAM format\n",
      " |      specification) as well as additional value type 'd' as\n",
      " |      implemented in htslib.\n",
      " |      \n",
      " |      Parameters\n",
      " |      ----------\n",
      " |      \n",
      " |      tag :\n",
      " |          data tag.\n",
      " |      \n",
      " |      with_value_type : Optional[bool]\n",
      " |          if set to True, the return value is a tuple of (tag value, type code).\n",
      " |          (default False)\n",
      " |      \n",
      " |      Returns\n",
      " |      -------\n",
      " |      \n",
      " |      A python object with the value of the `tag`. The type of the\n",
      " |      object depends on the data type in the data record.\n",
      " |      \n",
      " |      Raises\n",
      " |      ------\n",
      " |      \n",
      " |      KeyError\n",
      " |          If `tag` is not present, a KeyError is raised.\n",
      " |  \n",
      " |  get_tags(...)\n",
      " |      AlignedSegment.get_tags(self, with_value_type=False)\n",
      " |      the fields in the optional aligment section.\n",
      " |      \n",
      " |              Returns a list of all fields in the optional\n",
      " |              alignment section. Values are converted to appropriate python\n",
      " |              values. For example:\n",
      " |      \n",
      " |              [(NM, 2), (RG, \"GJP00TM04\")]\n",
      " |      \n",
      " |              If *with_value_type* is set, the value type as encode in\n",
      " |              the AlignedSegment record will be returned as well:\n",
      " |      \n",
      " |              [(NM, 2, \"i\"), (RG, \"GJP00TM04\", \"Z\")]\n",
      " |      \n",
      " |              This method will convert all values in the optional alignment\n",
      " |              section. When getting only one or few tags, please see\n",
      " |              :meth:`get_tag` for a quicker way to achieve this.\n",
      " |  \n",
      " |  has_tag(...)\n",
      " |      AlignedSegment.has_tag(self, tag)\n",
      " |      returns true if the optional alignment section\n",
      " |              contains a given *tag*.\n",
      " |  \n",
      " |  infer_query_length(...)\n",
      " |      AlignedSegment.infer_query_length(self, always=False)\n",
      " |      infer query length from CIGAR alignment.\n",
      " |      \n",
      " |              This method deduces the query length from the CIGAR alignment\n",
      " |              but does not include hard-clipped bases.\n",
      " |      \n",
      " |              Returns None if CIGAR alignment is not present.\n",
      " |      \n",
      " |              If *always* is set to True, `infer_read_length` is used instead.\n",
      " |              This is deprecated and only present for backward compatibility.\n",
      " |  \n",
      " |  infer_read_length(...)\n",
      " |      AlignedSegment.infer_read_length(self)\n",
      " |      infer read length from CIGAR alignment.\n",
      " |      \n",
      " |              This method deduces the read length from the CIGAR alignment\n",
      " |              including hard-clipped bases.\n",
      " |      \n",
      " |              Returns None if CIGAR alignment is not present.\n",
      " |  \n",
      " |  opt(...)\n",
      " |      AlignedSegment.opt(self, tag)\n",
      " |      deprecated, use get_tag() instead\n",
      " |  \n",
      " |  overlap(...)\n",
      " |      AlignedSegment.overlap(self)\n",
      " |      deprecated, use get_overlap() instead\n",
      " |  \n",
      " |  setTag(...)\n",
      " |      AlignedSegment.setTag(self, tag, value, value_type=None, replace=True)\n",
      " |      deprecated, use set_tag() instead\n",
      " |  \n",
      " |  set_tag(...)\n",
      " |      AlignedSegment.set_tag(self, tag, value, value_type=None, replace=True)\n",
      " |      sets a particular field *tag* to *value* in the optional alignment\n",
      " |              section.\n",
      " |      \n",
      " |              *value_type* describes the type of *value* that is to entered\n",
      " |              into the alignment record. It can be set explicitly to one of\n",
      " |              the valid one-letter type codes. If unset, an appropriate type\n",
      " |              will be chosen automatically based on the python type of\n",
      " |              *value*.\n",
      " |      \n",
      " |              An existing value of the same *tag* will be overwritten unless\n",
      " |              *replace* is set to False. This is usually not recommened as a\n",
      " |              tag may only appear once in the optional alignment section.\n",
      " |      \n",
      " |              If *value* is None, the tag will be deleted.\n",
      " |      \n",
      " |              This method accepts valid SAM specification value types, which\n",
      " |              are::\n",
      " |              \n",
      " |                 A: printable char\n",
      " |                 i: signed int\n",
      " |                 f: float\n",
      " |                 Z: printable string\n",
      " |                 H: Byte array in hex format\n",
      " |                 B: Integer or numeric array\n",
      " |      \n",
      " |              Additionally, it will accept the integer BAM types ('cCsSI')\n",
      " |      \n",
      " |              For htslib compatibility, 'a' is synonymous with 'A' and the\n",
      " |              method accepts a 'd' type code for a double precision float.\n",
      " |      \n",
      " |              When deducing the type code by the python type of *value*, the\n",
      " |              following mapping is applied::\n",
      " |              \n",
      " |                  i: python int\n",
      " |                  f: python float\n",
      " |                  Z: python str or bytes\n",
      " |                  B: python array.array, list or tuple\n",
      " |                  \n",
      " |              Note that a single character string will be output as 'Z' and\n",
      " |              not 'A' as the former is the more general type.\n",
      " |  \n",
      " |  set_tags(...)\n",
      " |      AlignedSegment.set_tags(self, tags)\n",
      " |      sets the fields in the optional alignment section with\n",
      " |              a list of (tag, value) tuples.\n",
      " |      \n",
      " |              The :term:`value type` of the values is determined from the\n",
      " |              python type. Optionally, a type may be given explicitly as\n",
      " |              a third value in the tuple, For example:\n",
      " |      \n",
      " |              x.set_tags([(NM, 2, \"i\"), (RG, \"GJP00TM04\", \"Z\")]\n",
      " |      \n",
      " |              This method will not enforce the rule that the same tag may appear\n",
      " |              only once in the optional alignment section.\n",
      " |  \n",
      " |  to_dict(...)\n",
      " |      AlignedSegment.to_dict(self)\n",
      " |      returns a json representation of the aligned segment.\n",
      " |      \n",
      " |              Field names are abbreviated versions of the class attributes.\n",
      " |  \n",
      " |  to_string(...)\n",
      " |      AlignedSegment.to_string(self)\n",
      " |      returns a string representation of the aligned segment.\n",
      " |      \n",
      " |              The output format is valid SAM format if a header is associated\n",
      " |              with the AlignedSegment.\n",
      " |  \n",
      " |  tostring(...)\n",
      " |      AlignedSegment.tostring(self, htsfile=None)\n",
      " |      deprecated, use :meth:`to_string()` instead.\n",
      " |      \n",
      " |              Parameters\n",
      " |              ----------\n",
      " |      \n",
      " |              htsfile -- (deprecated) AlignmentFile object to map numerical\n",
      " |                         identifiers to chromosome names. This parameter is present\n",
      " |                         for backwards compatibility and ignored.\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Class methods defined here:\n",
      " |  \n",
      " |  from_dict(...) from builtins.type\n",
      " |      AlignedSegment.from_dict(type cls, sam_dict, AlignmentHeader header)\n",
      " |      parses a dictionary representation of the aligned segment.\n",
      " |      \n",
      " |              Parameters\n",
      " |              ----------\n",
      " |              sam_dict -- dictionary of alignment values, keys corresponding to output from\n",
      " |                          :meth:`todict()`.\n",
      " |  \n",
      " |  fromstring(...) from builtins.type\n",
      " |      AlignedSegment.fromstring(type cls, sam, AlignmentHeader header)\n",
      " |      parses a string representation of the aligned segment.\n",
      " |      \n",
      " |              The input format should be valid SAM format.\n",
      " |      \n",
      " |              Parameters\n",
      " |              ----------\n",
      " |              sam -- :term:`SAM` formatted string\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Static methods defined here:\n",
      " |  \n",
      " |  __new__(*args, **kwargs) from builtins.type\n",
      " |      Create and return a new object.  See help(type) for accurate signature.\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Data descriptors defined here:\n",
      " |  \n",
      " |  aend\n",
      " |      deprecated, reference_end instead\n",
      " |  \n",
      " |  alen\n",
      " |      deprecated, reference_length instead\n",
      " |  \n",
      " |  aligned_pairs\n",
      " |      deprecated, use get_aligned_pairs() instead\n",
      " |  \n",
      " |  bin\n",
      " |      properties bin\n",
      " |  \n",
      " |  blocks\n",
      " |      deprecated, use get_blocks() instead\n",
      " |  \n",
      " |  cigar\n",
      " |      deprecated, use cigartuples instead\n",
      " |  \n",
      " |  cigarstring\n",
      " |      the :term:`cigar` alignment as a string.\n",
      " |      \n",
      " |      The cigar string is a string of alternating integers\n",
      " |      and characters denoting the length and the type of\n",
      " |      an operation.\n",
      " |      \n",
      " |      .. note::\n",
      " |          The order length,operation is specified in the\n",
      " |          SAM format. It is different from the order of\n",
      " |          the :attr:`cigar` property.\n",
      " |      \n",
      " |      Returns None if not present.\n",
      " |      \n",
      " |      To unset the cigarstring, assign None or the\n",
      " |      empty string.\n",
      " |  \n",
      " |  cigartuples\n",
      " |      the :term:`cigar` alignment. The alignment\n",
      " |      is returned as a list of tuples of (operation, length).\n",
      " |      \n",
      " |      If the alignment is not present, None is returned.\n",
      " |      \n",
      " |      The operations are:\n",
      " |      \n",
      " |      +-----+--------------+-----+\n",
      " |      |M    |BAM_CMATCH    |0    |\n",
      " |      +-----+--------------+-----+\n",
      " |      |I    |BAM_CINS      |1    |\n",
      " |      +-----+--------------+-----+\n",
      " |      |D    |BAM_CDEL      |2    |\n",
      " |      +-----+--------------+-----+\n",
      " |      |N    |BAM_CREF_SKIP |3    |\n",
      " |      +-----+--------------+-----+\n",
      " |      |S    |BAM_CSOFT_CLIP|4    |\n",
      " |      +-----+--------------+-----+\n",
      " |      |H    |BAM_CHARD_CLIP|5    |\n",
      " |      +-----+--------------+-----+\n",
      " |      |P    |BAM_CPAD      |6    |\n",
      " |      +-----+--------------+-----+\n",
      " |      |=    |BAM_CEQUAL    |7    |\n",
      " |      +-----+--------------+-----+\n",
      " |      |X    |BAM_CDIFF     |8    |\n",
      " |      +-----+--------------+-----+\n",
      " |      |B    |BAM_CBACK     |9    |\n",
      " |      +-----+--------------+-----+\n",
      " |      \n",
      " |      .. note::\n",
      " |          The output is a list of (operation, length) tuples, such as\n",
      " |          ``[(0, 30)]``.\n",
      " |          This is different from the SAM specification and\n",
      " |          the :attr:`cigarstring` property, which uses a\n",
      " |          (length, operation) order, for example: ``30M``.\n",
      " |      \n",
      " |      To unset the cigar property, assign an empty list\n",
      " |      or None.\n",
      " |  \n",
      " |  flag\n",
      " |      properties flag\n",
      " |  \n",
      " |  header\n",
      " |  \n",
      " |  inferred_length\n",
      " |      deprecated, use infer_query_length() instead\n",
      " |  \n",
      " |  is_duplicate\n",
      " |      true if optical or PCR duplicate\n",
      " |  \n",
      " |  is_paired\n",
      " |      true if read is paired in sequencing\n",
      " |  \n",
      " |  is_proper_pair\n",
      " |      true if read is mapped in a proper pair\n",
      " |  \n",
      " |  is_qcfail\n",
      " |      true if QC failure\n",
      " |  \n",
      " |  is_read1\n",
      " |      true if this is read1\n",
      " |  \n",
      " |  is_read2\n",
      " |      true if this is read2\n",
      " |  \n",
      " |  is_reverse\n",
      " |      true if read is mapped to reverse strand\n",
      " |  \n",
      " |  is_secondary\n",
      " |      true if not primary alignment\n",
      " |  \n",
      " |  is_supplementary\n",
      " |      true if this is a supplementary alignment\n",
      " |  \n",
      " |  is_unmapped\n",
      " |      true if read itself is unmapped\n",
      " |  \n",
      " |  isize\n",
      " |      deprecated, use template_length instead\n",
      " |  \n",
      " |  mapping_quality\n",
      " |      mapping quality\n",
      " |  \n",
      " |  mapq\n",
      " |      deprecated, use mapping_quality instead\n",
      " |  \n",
      " |  mate_is_reverse\n",
      " |      true is read is mapped to reverse strand\n",
      " |  \n",
      " |  mate_is_unmapped\n",
      " |      true if the mate is unmapped\n",
      " |  \n",
      " |  mpos\n",
      " |      deprecated, use next_reference_start instead\n",
      " |  \n",
      " |  mrnm\n",
      " |      deprecated, use next_reference_id instead\n",
      " |  \n",
      " |  next_reference_id\n",
      " |      the :term:`reference` id of the mate/next read.\n",
      " |  \n",
      " |  next_reference_name\n",
      " |      :term:`reference` name of the mate/next read (None if no\n",
      " |      AlignmentFile is associated)\n",
      " |  \n",
      " |  next_reference_start\n",
      " |      the position of the mate/next read.\n",
      " |  \n",
      " |  pnext\n",
      " |      deprecated, use next_reference_start instead\n",
      " |  \n",
      " |  pos\n",
      " |      deprecated, use reference_start instead\n",
      " |  \n",
      " |  positions\n",
      " |      deprecated, use get_reference_positions() instead\n",
      " |  \n",
      " |  qend\n",
      " |      deprecated, use query_alignment_end instead\n",
      " |  \n",
      " |  qlen\n",
      " |      deprecated, use query_alignment_length instead\n",
      " |  \n",
      " |  qname\n",
      " |      deprecated, use query_name instead\n",
      " |  \n",
      " |  qqual\n",
      " |      deprecated, query_alignment_qualities instead\n",
      " |  \n",
      " |  qstart\n",
      " |      deprecated, use query_alignment_start instead\n",
      " |  \n",
      " |  qual\n",
      " |      deprecated, query_qualities instead\n",
      " |  \n",
      " |  query\n",
      " |      deprecated, query_alignment_sequence instead\n",
      " |  \n",
      " |  query_alignment_end\n",
      " |      end index of the aligned query portion of the sequence (0-based,\n",
      " |      exclusive)\n",
      " |      \n",
      " |      This the index just past the last base in :attr:`seq` that is not\n",
      " |      soft-clipped.\n",
      " |  \n",
      " |  query_alignment_length\n",
      " |      length of the aligned query sequence.\n",
      " |      \n",
      " |      This is equal to :attr:`qend` - :attr:`qstart`\n",
      " |  \n",
      " |  query_alignment_qualities\n",
      " |      aligned query sequence quality values (None if not present). These\n",
      " |      are the quality values that correspond to :attr:`query`, that\n",
      " |      is, they exclude qualities of :term:`soft clipped` bases. This\n",
      " |      is equal to ``qual[qstart:qend]``.\n",
      " |      \n",
      " |      Quality scores are returned as a python array of unsigned\n",
      " |      chars. Note that this is not the ASCII-encoded value typically\n",
      " |      seen in FASTQ or SAM formatted files. Thus, no offset of 33\n",
      " |      needs to be subtracted.\n",
      " |      \n",
      " |      This property is read-only.\n",
      " |  \n",
      " |  query_alignment_sequence\n",
      " |      aligned portion of the read.\n",
      " |      \n",
      " |      This is a substring of :attr:`seq` that excludes flanking\n",
      " |      bases that were :term:`soft clipped` (None if not present). It\n",
      " |      is equal to ``seq[qstart:qend]``.\n",
      " |      \n",
      " |      SAM/BAM files may include extra flanking bases that are not\n",
      " |      part of the alignment.  These bases may be the result of the\n",
      " |      Smith-Waterman or other algorithms, which may not require\n",
      " |      alignments that begin at the first residue or end at the last.\n",
      " |      In addition, extra sequencing adapters, multiplex identifiers,\n",
      " |      and low-quality bases that were not considered for alignment\n",
      " |      may have been retained.\n",
      " |  \n",
      " |  query_alignment_start\n",
      " |      start index of the aligned query portion of the sequence (0-based,\n",
      " |      inclusive).\n",
      " |      \n",
      " |      This the index of the first base in :attr:`seq` that is not\n",
      " |      soft-clipped.\n",
      " |  \n",
      " |  query_length\n",
      " |      the length of the query/read.\n",
      " |      \n",
      " |      This value corresponds to the length of the sequence supplied\n",
      " |      in the BAM/SAM file. The length of a query is 0 if there is no\n",
      " |      sequence in the BAM/SAM file. In those cases, the read length\n",
      " |      can be inferred from the CIGAR alignment, see\n",
      " |      :meth:`pysam.AlignedSegment.infer_query_length`.\n",
      " |      \n",
      " |      The length includes soft-clipped bases and is equal to\n",
      " |      ``len(query_sequence)``.\n",
      " |      \n",
      " |      This property is read-only but can be set by providing a\n",
      " |      sequence.\n",
      " |      \n",
      " |      Returns 0 if not available.\n",
      " |  \n",
      " |  query_name\n",
      " |      the query template name (None if not present)\n",
      " |  \n",
      " |  query_qualities\n",
      " |      read sequence base qualities, including :term:`soft\n",
      " |      clipped` bases (None if not present).\n",
      " |      \n",
      " |      Quality scores are returned as a python array of unsigned\n",
      " |      chars. Note that this is not the ASCII-encoded value typically\n",
      " |      seen in FASTQ or SAM formatted files. Thus, no offset of 33\n",
      " |      needs to be subtracted.\n",
      " |      \n",
      " |      Note that to set quality scores the sequence has to be set\n",
      " |      beforehand as this will determine the expected length of the\n",
      " |      quality score array.\n",
      " |      \n",
      " |      This method raises a ValueError if the length of the\n",
      " |      quality scores and the sequence are not the same.\n",
      " |  \n",
      " |  query_sequence\n",
      " |      read sequence bases, including :term:`soft clipped` bases\n",
      " |      (None if not present).\n",
      " |      \n",
      " |      Note that assigning to seq will invalidate any quality scores.\n",
      " |      Thus, to in-place edit the sequence and quality scores, copies of\n",
      " |      the quality scores need to be taken. Consider trimming for example::\n",
      " |      \n",
      " |         q = read.query_qualities\n",
      " |         read.query_squence = read.query_sequence[5:10]\n",
      " |         read.query_qualities = q[5:10]\n",
      " |      \n",
      " |      The sequence is returned as it is stored in the BAM file. Some mappers\n",
      " |      might have stored a reverse complement of the original read\n",
      " |      sequence.\n",
      " |  \n",
      " |  reference_end\n",
      " |      aligned reference position of the read on the reference genome.\n",
      " |      \n",
      " |      reference_end points to one past the last aligned residue.\n",
      " |      Returns None if not available (read is unmapped or no cigar\n",
      " |      alignment present).\n",
      " |  \n",
      " |  reference_id\n",
      " |      :term:`reference` ID\n",
      " |      \n",
      " |      .. note::\n",
      " |      \n",
      " |          This field contains the index of the reference sequence in\n",
      " |          the sequence dictionary. To obtain the name of the\n",
      " |          reference sequence, use :meth:`get_reference_name()`\n",
      " |  \n",
      " |  reference_length\n",
      " |      aligned length of the read on the reference genome.\n",
      " |      \n",
      " |      This is equal to `aend - pos`. Returns None if not available.\n",
      " |  \n",
      " |  reference_name\n",
      " |      :term:`reference` name\n",
      " |  \n",
      " |  reference_start\n",
      " |      0-based leftmost coordinate\n",
      " |  \n",
      " |  rlen\n",
      " |      deprecated, query_length instead\n",
      " |  \n",
      " |  rname\n",
      " |      deprecated, use reference_id instead\n",
      " |  \n",
      " |  rnext\n",
      " |      deprecated, use next_reference_id instead\n",
      " |  \n",
      " |  seq\n",
      " |      deprecated, use query_sequence instead\n",
      " |  \n",
      " |  tags\n",
      " |      deprecated, use get_tags() instead\n",
      " |  \n",
      " |  template_length\n",
      " |      the observed query template length\n",
      " |  \n",
      " |  tid\n",
      " |      deprecated, use reference_id instead\n",
      " |  \n",
      " |  tlen\n",
      " |      deprecated, use template_length instead\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Data and other attributes defined here:\n",
      " |  \n",
      " |  __pyx_vtable__ = <capsule object NULL>\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "<pysam.libcalignmentfile.AlignmentHeader at 0x108fa88f0>"
      ]
     },
     "execution_count": 472,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bam_file = \"data/IL27/flye.bam\"\n",
    "bam = pysam.AlignmentFile(bam_file, 'rb')\n",
    "contigs = [contig for contig in bam.fetch()]\n",
    "help(contigs[1])\n",
    "contigs[1].header"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 477,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[('NM', 1656),\n",
       " ('ms', 460610),\n",
       " ('AS', 460632),\n",
       " ('nn', 0),\n",
       " ('tp', 'P'),\n",
       " ('cm', 37508),\n",
       " ('s1', 210541),\n",
       " ('s2', 788),\n",
       " ('de', 0.0015999999595806003)]"
      ]
     },
     "execution_count": 477,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "contig = contigs[1]\n",
    "contig.get_tags()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
